### PR TITLE
fix optional cram support in modules and migrate to nf-test

### PR DESCRIPTION
**Pull request by tomiles**
- Make reference genome fasta optional for cram support and split outputs for cram versus bam/sam where appropriate. 
- Migrate modules to nf-test

## PR checklist
- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/picard/markduplicates/main.nf [hunk 1]
```diff
@@ -24,23 +25,27 @@ process PICARD_MARKDUPLICATES {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def suffix = task.ext.suffix    ?: "${reads.getExtension()}"
+    def reference = fasta ? "--REFERENCE_SEQUENCE ${fasta}" : ""
+    def create_index = ( suffix == "bam" )? "--CREATE_INDEX" : ""
```
Comment (matthdsm) [inline modules/nf-core/picard/markduplicates/main.nf:30]: `--create_index` shouldn't be added in the module, but added by the user through `ext.args`

**Review by matthdsm — CHANGES_REQUESTED** [review]

File: modules/nf-core/picard/markduplicates/main.nf (hunk 1 above)
Comment (tomiles) [reply modules/nf-core/picard/markduplicates/main.nf]: removed it

File: modules/nf-core/picard/addorreplacereadgroups/main.nf [hunk 2]
```diff
@@ -24,26 +24,24 @@ process PICARD_ADDORREPLACEREADGROUPS {
     script:
     def args = task.ext.args        ?: ''
     def prefix = task.ext.prefix    ?: "${meta.id}"
-    def suffix = task.ext.suffix    ?: "${bam.getExtension()}"
+    def suffix = task.ext.suffix    ?: "${reads.getExtension()}"
     def reference = fasta ? "--REFERENCE_SEQUENCE ${fasta}" : ""
-    def create_index = ( suffix == "bam" )? "--CREATE_INDEX" : ""
```
Comment (maxulysse) [inline modules/nf-core/picard/addorreplacereadgroups/main.nf:29]: so no creating index anymore?

Comment (matthdsm) [reply modules/nf-core/picard/addorreplacereadgroups/main.nf:29]: Yes to creating index, but it's a flag that should be set by the user in `ext.args`

Comment (tomiles) [reply modules/nf-core/picard/addorreplacereadgroups/main.nf:29]: I added `--CREATE_INDEX` to `ext.args` in one of the test configs just so we keep testing index generation works if configured and that the bai is captured in the proper output

**Pull request merged**